crates/cow-amm/src/registry.rs

```diff
@@ -0,0 +1,171 @@
+use {
+    crate::{factories::Deployment, Amm},
+    anyhow::Context,
+    contracts::CowAmmLegacyHelper,
+    ethcontract::Address,
+    ethrpc::{
+        current_block::{BlockRetrieving, CurrentBlockStream, RangeInclusive},
+        Web3,
+    },
+    shared::{
+        event_handling::{EventHandler, EventRetrieving, EventStoring},
+        maintenance::{Maintaining, ServiceMaintenance},
+    },
+    std::{
+        any::TypeId,
+        collections::{BTreeMap, HashMap},
+        sync::Arc,
+    },
+    tokio::sync::{Mutex, RwLock},
+};
+
+/// CoW AMM indexer which stores events in-memory.
+#[derive(Clone)]
+pub struct Registry {
+    /// Store indexed data associated to the indexed events type id.
+    /// That type erasure allows us to index multiple concrete contracts
+    /// in a single Registry to make for a nicer user facing API.
+    storage: Arc<RwLock<HashMap<TypeId, Storage>>>,
+    web3: Web3,
+    current_block_stream: CurrentBlockStream,
+}
+
+impl Registry {
+    pub fn new(web3: Web3, current_block_stream: CurrentBlockStream) -> Self {
+        Self {
+            storage: Default::default(),
+            web3,
+            current_block_stream,
+        }
+    }
+
+    pub async fn add_listener<C>(&self, first_block: u64, contract: C, helper_contract: Address)
```

Comment:
Conceptual question, so it is not possible for new contracts to be detected with a different block stream than the one we use right?

Reply:
Theoretically we could use a new `CurrentBlockStream` for each contract we index. That component is just there to give the code a "pulse" (to let it know when a new block happened so we can poll for new data on it).
The nice thing about using the same instance everywhere is that it reduces race conditions because all components should learn about new block roughly at the same time (some variance can't be avoided to the tokio runtime).

Reply:
I understand! thanks a lot for the explanation! :raised_hands: